### PR TITLE
feature(signl4): onboarding signl4 source

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

> onboarding signl4 source

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/sources/signl4/transform.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  return message;
+}
+
+function process(event) {
+  let message = new Message(`Signl4`);
+
+  // we are setting event type as track always
+  const eventType = "track";
+
+  message.setEventType(eventType);
```
Comment (ItsSudip) [inline v0/sources/signl4/transform.js:42]:
```suggestion
  message.setEventType("track");
```

File: v0/sources/signl4/transform.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+  message.setPropertiesV2(event, mapping);
+
+  // setting event Name
+  switch (event.eventType) {
+    case 200:
+      message.setEventName("New Signl Created");
```
Comment (ItsSudip) [inline v0/sources/signl4/transform.js:49]: Let's rename signl to alert.

File: v0/sources/signl4/transform.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+const Message = require("../message");
+
+// import mapping json using JSON.parse to preserve object key order
+const mapping = JSON.parse(
+  fs.readFileSync(path.resolve(__dirname, "./mapping.json"), "utf-8")
+);
+
+function settingProperties(event, message) {
+  const flatEvent = event;
+  const messageReplica = message;
```
Comment (ItsSudip) [inline v0/sources/signl4/transform.js:13]: No need to use another variable. We can directly operate on `message` and `event`.

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: v0/sources/signl4/transform.js (hunk 1 above)
Comment (ujjwal-ab) [reply v0/sources/signl4/transform.js]: Done

File: v0/sources/signl4/transform.js (hunk 2 above)
Comment (ujjwal-ab) [reply v0/sources/signl4/transform.js]: Done

File: v0/sources/signl4/transform.js (hunk 3 above)
Comment (ujjwal-ab) [reply v0/sources/signl4/transform.js]: Removed the flatEvent one but for the message, eslint was disallowing the assignment to the function parameter. That's why create a replica for this.

File: v0/sources/signl4/transform.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    case 301:
+      message.setEventName("Duty Period Ended");
+      break;
+    case 302:
+      message.setEventName("Somebody Punched-In");
+      break;
+    case 303:
+      message.setEventName("Somebody Punched-Out");
+      break;
+    default:
```
Comment (ItsSudip) [inline v0/sources/signl4/transform.js:71]: We should hadle a default case as well. Something like `Alert Updated`.

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: v0/sources/signl4/transform.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  // Here, we are checking for the test event to discard them
+  if (event.eventType === 1) {
+    return {
+      outputToSource: {
+        body: Buffer.from(JSON.stringify(event)).toString("base64"),
+        contentType: "application/json"
+      },
+      statusCode: 200
+    };
+  }
```
Comment (ItsSudip) [inline v0/sources/signl4/transform.js:112]: Move this in front of the code.

Comment (ujjwal-ab) [reply v0/sources/signl4/transform.js]: Done

**Review by ItsSudip — APPROVED** [review]

**Review by sauravlal-rudderstack — APPROVED** [review]

**Pull request merged**